### ISSUE TITLE
scheduling pipelines at a given revision of a material fails on phabricator-type staging repos

### ISSUE THREAD

**Issue by drmuey** [gocd/gocd#1576]
Some development systems have the idea of a staging repo that pushes local changes to build/test as a single tagged SHA[1]. The problem is that Go can’t see those SHAs when scheduling a specific one.

For example, given this:

```
$ arc diff
…
* [new tag]         8d09dc41096c9e5a26a609cb3d0a064f7d5f4ead -> phabricator/diff/72576
STAGING PUSHED  Pushed a copy of the changes to tag "phabricator/diff/72576" in the staging area.
Updated an existing Differential revision:
…
```

Now I'd like to start a build of 8d09dc41096c9e5a26a609cb3d0a064f7d5f4ead but when you:

`POST http:/…/go/api/pipelines/phabricator-build/schedule?materials[test_staging]=8d09dc41096c9e5a26a609cb3d0a064f7d5f4ead`

you get a 202 w/ “[Request to schedule pipeline](https://github.com/gocd/gocd/blob/ad8fc27d552f038fc8609c27cd0d6e4eadd90739/server/src/com/thoughtworks/go/server/scheduling/BuildCauseProducerService.java#L300) phabricator-build accepted” but it never starts. Checking the log you see:

```
2015-10-20 12:24:37,564 ERROR [60@MessageListener for MaterialUpdateStatusNotifier] BuildCauseProducerService:244 - Error while scheduling pipeline: phabricator-build
java.lang.RuntimeException: Unable to find revision [8d09dc41096c9e5a26a609cb3d0a064f7d5f4ead] for material [GitMaterial{url=http://….git, branch='master', submoduleFolder='null'}]
        at com.thoughtworks.go.server.materials.MaterialChecker.findSpecificRevision(MaterialChecker.java:52)
```

~~If it did `git fetch --tags` first it’d have 8d09dc41096c9e5a26a609cb3d0a064f7d5f4ead~~ **Update**:

I built a patched installer to test that and it turns out that it makes no difference. With or without the original patch git has access to the SHA and the TAG during a build. It is when scheduling a pipeline and specifying a material revision that it falls apart. The [failure to find the revision](https://github.com/gocd/gocd/blob/ad8fc27d552f038fc8609c27cd0d6e4eadd90739/server/src/com/thoughtworks/go/server/materials/MaterialChecker.java#L52) comes, not from the repo, but from [an SQL query](https://github.com/gocd/gocd/blob/f4df77cc1f6fe37bd64f27043b3f73dcb915dc2e/server/src/com/thoughtworks/go/server/persistence/MaterialRepository.java#L875). It is [as if the internal SQL is out of date](https://github.com/gocd/gocd/blob/f4df77cc1f6fe37bd64f27043b3f73dcb915dc2e/server/src/com/thoughtworks/go/server/materials/MaterialDatabaseUpdater.java#L138) w/ the repo.

I have such a [repo setup for easily reproducing](#issuecomment-166900185) the problem.

**Footnote**:

[1] The relevant bits from the phabricator/harbormaster docs:

> With staging areas, arc pushes a copy of the local changes somewhere as a side effect of running arc diff. In Git, it pushes changes to a tag like phabricator/diff/123 in a designated remote.
> 
> The build system can then interact with this copy using normal VCS commands. 
> …
> Manual staging areas work like this:
> - You configure a staging area for the repository you want to be able to run builds for. A staging area is just a remote repository that you're designating for temporary storage.
> - Once a staging area is configured, arc diff will automatically push a copy of the changes to the staging area as a side effect when creating and updating reviews.
> - Your build system can pull changes directly from the configured staging area.

**Comment by drmuey** [comment]
Would it be as simple as this?

```
- String[] args = new String[]{"pull"};
+ String[] args = new String[]{"pull --tags"};
```
at 
https://github.com/gocd/gocd/blob/ad8fc27d552f038fc8609c27cd0d6e4eadd90739/common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java#L253

**Comment by drmuey** [comment]
Well, probably more like this:

```
- String[] args = new String[]{"pull"};
- CommandLine gitCommit = git().withArgs(args).withWorkingDir(workingDir);
+ CommandLine gitCommit = git().withArgs("pull", "--tags").withWorkingDir(workingDir);
```

but still, very simple and low impact no?

**Comment by arvindsv** [comment]
:) It sounds simple. I don't know what else it will impact. But, if you're game, you can make a change in a fork and [build an installer](https://hub.docker.com/r/gocd/gocd-build-installer/) using it and see if it works. That would be enormously helpful, and will be a strong case to make that change. Some tests will need to be written, of course, but the ground work will be done.

**Comment by drmuey** [comment]
@arvindsv sounds good thanks, I’ll update the issue w/ details as we’re able to do it

**Comment by drmuey** [comment]
FWiW:

*Fork*: https://github.com/drmuey/gocd.git 
*Branch*: Issue1576

**Comment by ketan** [comment]
Can you submit a PR (even if it is not complete yet) so someone can review and provide feedback?

**Comment by drmuey** [comment]
@ketan here is the pull request you asked for thanks! I am in the process of building the installer to test (docker/VM issues … blarg)

#1619

**Comment by drmuey** [comment]
For anyone playing along at home, here is the current state of my testing:

`docker run -it -v `pwd`/installers:/installers -e REPO=https://github.com/drmuey/gocd.git -e BRANCH=Issue1576 gocd/gocd-build-installer rpm`

fails w/:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.3:compile (default-compile) on project common: Compilation failure: Compilation failure:
[ERROR] /home/gocd/common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java:[253,64] cannot find symbol
[ERROR] symbol:   variable args
[ERROR] location: class com.thoughtworks.go.domain.materials.git.GitCommand
[ERROR] /home/gocd/common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java:[253,33] method git in class com.thoughtworks.go.domain.materials.git.GitCommand cannot be applied to given types;
[ERROR] required: no arguments
[ERROR] found: java.lang.String,java.lang.String
[ERROR] reason: actual and formal argument lists differ in length
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :common
Buildr aborted!
RuntimeError : Failed to run: mvn install -DskipTests
```

**Comment by arvindsv** [comment]
@drmuey: [This](https://github.com/drmuey/gocd/commit/57c32b25a9a30e69adc7bfd10dda9db4671b5c8a#diff-97dd991971202dc0de417dbb4271aa82R253) is your change, right? Shouldn't it be something like this:

```
String[] args = new String[]{"pull", "--tags"};
CommandLine gitCommit = git().withArgs(args).withWorkingDir(workingDir)
```

Your change doesn't compile for me, in IntelliJ. That's what the failure is showing as well.

**Comment by drmuey** [comment]
and …  `docker run -it --rm=true -v `pwd`/out:/installers gocd/gocd-build-installer rpm` works fine, derp!

Testing w/ `--rm=true` ATM

**Comment by arvindsv** [comment]
The installer should be in the directory: "out" and you can try it out. Even with `--rm=true`.

**Comment by drmuey** [comment]
@arvindsv 
thanks, I'll update the change and try again. I thought it would work since other ones do strings directly.

**Comment by drmuey** [comment]
@arvindsv ah actually I had the args in the wrong method, derp! updated branch and am running it now, thanks for mental handle jiggling ;)

**Comment by drmuey** [comment]
#1619 @ bee6d29 via `docker run -it --rm=true -v `pwd`/installers:/installers -e REPO=https://github.com/drmuey/gocd.git -e BRANCH=Issue1576 gocd/gocd-build-installer rpm osx`

* RPM
** [Go Server](https://drive.google.com/file/d/0BzDv1HF2jIW7TVA0UER3T2tGSkU/view?usp=sharing)
** [Go Agent](https://drive.google.com/file/d/0BzDv1HF2jIW7aXBJZUtxMGhkMTA/view?usp=sharing)
* OSX 
** [Go Server](https://drive.google.com/file/d/0BzDv1HF2jIW7cXZhbThMYko4WFk/view?usp=sharing)
** [Go Agent](https://drive.google.com/file/d/0BzDv1HF2jIW7Qk51RHRwYndVM28/view?usp=sharing)

**Comment by drmuey** [comment]
Does not seem to do the trick :(

It should be noted that if I `git show TAG/IN/QUESTION` and `git show SHAofTAG` in a job it is able to see them both (before and after my change). It just the initial scheduling that can't see it (error below) as if it is `git checkout SHA` before the git pull.

```
015-12-14 07:22:22,648 ERROR [62@MessageListener for MaterialUpdateStatusNotifier] BuildCauseProducerService:244 - Error while scheduling pipeline: test_gh_1576
java.lang.RuntimeException: Unable to find revision [c16bb76e5520808cb56a9ef2ae89025435135c8f] for material [GitMaterial{url=http://127.0.0.1/derprepo/PhabricatorManualStagingArea-Admin.git, branch='master', submoduleFolder='null'}]
	at com.thoughtworks.go.server.materials.MaterialChecker.findSpecificRevision(MaterialChecker.java:52)
	at com.thoughtworks.go.server.materials.SpecificMaterialRevisionFactory.create(SpecificMaterialRevisionFactory.java:53)
	at com.thoughtworks.go.server.scheduling.BuildCauseProducerService.newProduceBuildCause(BuildCauseProducerService.java:175)
	at com.thoughtworks.go.server.scheduling.BuildCauseProducerService$WaitForPipelineMaterialUpdate.onMaterialUpdate(BuildCauseProducerService.java:323)
	at com.thoughtworks.go.server.materials.MaterialUpdateStatusNotifier.onMessage(MaterialUpdateStatusNotifier.java:55)
	at com.thoughtworks.go.server.materials.MaterialUpdateStatusNotifier.onMessage(MaterialUpdateStatusNotifier.java:29)
	at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.runImpl(JMSMessageListenerAdapter.java:65)
	at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.run(JMSMessageListenerAdapter.java:50)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by drmuey** [comment]
Ah, so the (failure to find the revision)[https://github.com/gocd/gocd/blob/ad8fc27d552f038fc8609c27cd0d6e4eadd90739/server/src/com/thoughtworks/go/server/materials/MaterialChecker.java#L52] comes, not from the repo, but from (an SQL query)[https://github.com/gocd/gocd/blob/f4df77cc1f6fe37bd64f27043b3f73dcb915dc2e/server/src/com/thoughtworks/go/server/persistence/MaterialRepository.java#L875].

So the problem seems to be that the embedded database is not up to date with the repo. Back to the drawing board …

**Comment by arvindsv** [comment]
I wonder if it makes sense to write a material plugin instead, which handles git materials like this. There's also an open issue about [handling shallow clones - #313](https://github.com/gocd/gocd/issues/313). I don't know. It'll have to be analyzed a bit. Especially how to handle tags themselves.

**Comment by donaldguy** [comment]
I for one suspect there could stand to be a general Phabricator diffs material plugin and might write one if I find the time

Another plugin I want and might write (for use on our master branch) is one that exposes the material label as the result of `git describe` rather than sha (while I'm wishlisting)

**Comment by drmuey** [comment]
> I wonder if it makes sense to write a material plugin instead, which handles git materials like this.

@arvindsv maybe, though it seems to not be a git/tag issue at all. The SHA is there, jobs have access to it. Its scheduling a build w/ a given SHA that fails, not because git can't find it but because the internal SQL command fails to find it. Like Go’s SQL is not up to date w/ the repo.

**Comment by drmuey** [comment]
> I for one suspect there could stand to be a general Phabricator diffs material plugin and might write one if I find the time

@donaldguy That’d be great ;) For now though it'd probably be much simpler just to get the scheduler to update the internal SQL with the latest material info before checking for the SHA

**Comment by drmuey** [comment]
Fro anyone following along at home: I updated the title and added an “Update” chunk to the original post to reflect the current state of the issue, thanks!

**Comment by drmuey** [comment]
New tak:

`git log` does not list the tags’ SHAs, `git log --tags` does. If the internal database is essentially populated from `git log`, then that is what needs changed and tested. Tracing that out now …

**Comment by drmuey** [comment]
The only two log functions probably won't be what needs done:
1. [get latest](https://github.com/gocd/gocd/blob/ad8fc27d552f038fc8609c27cd0d6e4eadd90739/common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java#L66) is likely only used when scheduling and not specifying a material revision
2. [get range](https://github.com/gocd/gocd/blob/ad8fc27d552f038fc8609c27cd0d6e4eadd90739/common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java#L70) when given one of these tag–revisions git log sees it just fine (w/ or w/out --tags works fine too, either way the range is a bit wonky)

Maybe it tries to check it out and does not like it because it’d be in a detached head state?

**Comment by arvindsv** [comment]
@drmuey: Sorry I missed your earlier message about Go's SQL. Phabricator (or whatever) would have done the equivalent of `git push --tags`, I guess. So, the tag is available at the remote and will be fetched when Go does a `git pull`. But ... I had assumed wrong: this is not about tags at all.

Your initial idea was about scheduling a build with: `8d09dc41096c9e5a26a609cb3d0a064f7d5f4ead`. That *should* work, because `8d09dc41096c9e5a26a609cb3d0a064f7d5f4ead` is a normal commit, right? If it is, then Go should see it as a modification.

If it is not, what is happening? Do you have any idea? How is phabricator and tags coming into this? If it's a commit, Go should see it. As soon as it sees it, and inserts a modification into the database, it should work.

Are you able to reproduce this without have phabricator in place. Maybe with a simple repo and some tags and commits? If you can do that, maybe I can take a look.

**Comment by drmuey** [comment]
@arvindsv thanks! I was just going to start sorting out a-repo-only way to reproduce, I'll update with details ASAP

**Comment by drmuey** [comment]
@arvindsv here you go! thanks!

TL;DR:

1. use [this repo](https://github.com/drmuey/gocd_issue1576_test) as a material.
2. Try to schedule a build for a tagged SHA (i.e. fd097731adf3aa10cc48065e8f429a085c675a23 or c7ede06ea700e447a6e239b829ff909b4e919dd6)

Details:

[Staging repo](https://github.com/drmuey/gocd_issue1576_test) setup, like normal:

```
• [dmuey@multivac:gocd_issue1576_test]$ g init .
Initialized empty Git repository in /Users/dmuey/git/.brooklyn/issue1576_test/.git/
• [dmuey@multivac:gocd_issue1576_test] (master)$ echo "hello world, I am a staging repo" > README.md
• [dmuey@multivac:gocd_issue1576_test] (master)$ g add README.md 
• [dmuey@multivac:gocd_issue1576_test] (master)$ g ci -m "Initial commit of staging repo"
[master (root-commit) 6134260] Initial commit of staging repo
 1 file changed, 1 insertion(+)
 create mode 100644 README.md
• [dmuey@multivac:gocd_issue1576_test] (master)$ g lg
6134260 Initial commit of staging repo
• [dmuey@multivac:gocd_issue1576_test] (master)$  g tag
• [dmuey@multivac:gocd_issue1576_test] (master)$ git remote add origin https://github.com/drmuey/gocd_issue1576_test.git
• [dmuey@multivac:gocd_issue1576_test] (master)$ git push -u origin master
Username for 'https://github.com': drmuey
Password for 'https://drmuey@github.com': 
Counting objects: 3, done.
Writing objects: 100% (3/3), 265 bytes | 0 bytes/s, done.
Total 3 (delta 0), reused 0 (delta 0)
To https://github.com/drmuey/gocd_issue1576_test.git
 * [new branch]      master -> master
Branch master set up to track remote branch master from origin by rebasing.
• [dmuey@multivac:gocd_issue1576_test] (master)$ 
```

Stage a change via tag (this is [how phabricator via arcanist does it](https://github.com/phacility/arcanist/blob/dae2f0073f019221839924b4b1431c6cba5dcdda/src/workflow/ArcanistDiffWorkflow.php#L2723)):

```
• [dmuey@multivac:gocd] (Issue1576_log_cmd)$ git push --no-verify -- git@github.com:drmuey/gocd_issue1576_test.git fd097731adf3aa10cc48065e8f429a085c675a23:refs/tags/test/tag/1st
Counting objects: 73259, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (27793/27793), done.
Writing objects: 100% (73259/73259), 219.64 MiB | 703.00 KiB/s, done.
Total 73259 (delta 36825), reused 72786 (delta 36352)
To git@github.com:drmuey/gocd_issue1576_test.git
 * [new tag]         fd097731adf3aa10cc48065e8f429a085c675a23 -> test/tag/1st
• [dmuey@multivac:gocd] (Issue1576_log_cmd)$ git push --no-verify -- git@github.com:drmuey/gocd_issue1576_test.git c7ede06ea700e447a6e239b829ff909b4e919dd6:refs/tags/test/tag/2nd
Total 0 (delta 0), reused 0 (delta 0)
To git@github.com:drmuey/gocd_issue1576_test.git
 * [new tag]         c7ede06ea700e447a6e239b829ff909b4e919dd6 -> test/tag/2nd
• [dmuey@multivac:gocd] (Issue1576_log_cmd)$ 
```

Tagged Change disposition in staging repo:

```
• [dmuey@multivac:gocd_issue1576_test] (master)$ g pull
Current branch master is up to date.
• [dmuey@multivac:gocd_issue1576_test] (master)$ g lg
6134260 Initial commit of staging repo
• [dmuey@multivac:gocd_issue1576_test] (master)$ g lg --tags
6134260 Initial commit of staging repo
• [dmuey@multivac:gocd_issue1576_test] (master)$ g pull --tags
remote: Counting objects: 73259, done.
remote: Compressing objects: 100% (27320/27320), done.
remote: Total 73259 (delta 36825), reused 73259 (delta 36825), pack-reused 0
Receiving objects: 100% (73259/73259), 219.63 MiB | 376.00 KiB/s, done.
Resolving deltas: 100% (36825/36825), done.
From https://github.com/drmuey/gocd_issue1576_test
 * [new tag]         test/tag/1st -> test/tag/1st
 * [new tag]         test/tag/2nd -> test/tag/2nd
Current branch master is up to date.
• [dmuey@multivac:gocd_issue1576_test] (master)$ g lg
6134260 Initial commit of staging repo
• [dmuey@multivac:gocd_issue1576_test] (master)$ g tag
test/tag/1st
test/tag/2nd
• [dmuey@multivac:gocd_issue1576_test] (master)$ g show test/tag/1st
commit fd097731adf3aa10cc48065e8f429a085c675a23
Author: Jyoti Singh <jyotisingh7@gmail.com>
Date:   Fri Dec 18 20:30:58 2015 +0530

    Update README.md

diff --git a/README.md b/README.md
index 326d53b..c80d7d3 100644
--- a/README.md
+++ b/README.md
@@ -17,7 +17,7 @@ See [#1104](https://github.com/gocd/gocd/issues/1104) for what's planned for the
 ## Contributing
 
 We encourage you to contribute to Go. For information on contributing to this project, please see our <a href=
-A lot of useful information like links to user documentation, design documentation, mailing lists etc can be f
+A lot of useful information like links to user documentation, design documentation, mailing lists etc. can be 
 
 ## License
 
• [dmuey@multivac:gocd_issue1576_test] (master)$ g show fd097731adf3aa10cc48065e8f429a085c675a23
commit fd097731adf3aa10cc48065e8f429a085c675a23
Author: Jyoti Singh <jyotisingh7@gmail.com>
Date:   Fri Dec 18 20:30:58 2015 +0530

    Update README.md

diff --git a/README.md b/README.md
index 326d53b..c80d7d3 100644
--- a/README.md
+++ b/README.md
@@ -17,7 +17,7 @@ See [#1104](https://github.com/gocd/gocd/issues/1104) for what's planned for the
 ## Contributing
 
 We encourage you to contribute to Go. For information on contributing to this project, please see our <a href=
-A lot of useful information like links to user documentation, design documentation, mailing lists etc can be f
+A lot of useful information like links to user documentation, design documentation, mailing lists etc. can be 
 
 ## License
 
• [dmuey@multivac:gocd_issue1576_test] (master)$ 

`g show test/tag/2nd` and `g show c7ede06ea700e447a6e239b829ff909b4e919dd6` also show that tagged change
```

**Comment by webframp** [comment]
Definitely interested, but not sure where to go next

**Comment by donaldguy** [comment]
So in the phabricator specific case I worked around kinda convoluted-ly by making a harbormaster build plan like

<img width="748" alt="screen shot 2016-08-25 at 10 02 19 pm" src="https://cloud.githubusercontent.com/assets/43136/17991762/a238ded4-6b0f-11e6-9c22-43fa7dc9737f.png">

so that it populates a branch

**Comment by donaldguy** [comment]
The steps of that maze by the way are:

1. create an almanac device, 
2. create an almanac network, 
3. create an alamanac port on the device associated with the network, 
4. add a public key to the almanac device, 
5. create an almanac service and bind it to the device
6. create a drydock host blueprint attached to the almanac service and (another new resource for) the private key for that public key from step 4, 
7. manually distribute that keypair to the host (in my case a new user on my phabricator server with the appropriate homedir /var/drydock) 
8. create a drydock working copy blueprint associated with the service resource
9. Approve the authorization of the working copy blueprint to use the host blueprint
10. Create a staging repo and add it to the main repo
11. Create a harbormaster build plan.
 - 10a. create a harbormaster build step to create a lease on the working copy drydock blueprint
 - 10b. approve the authorization of that build step
12. Create the build step that actually does stuff
13. Make sure that appropriate permissions and keys existed for the user to access the two repos (I made a new bot user with a new keypair)
14. Create a herald rule to execute the build plan on new diffs
15. configure go.cd to use the feature-branch plugin against the staging repo (populating master with a readme so that the first run doesn't fail)

**Comment by drmuey** [comment]
@donaldguy thanks! Will see about giving it a try soon. @webframp not sure if you saw the last two posts so pinging you :)

**Comment by webframp** [comment]
Great set of steps @donaldguy thanks! Definitely a maze to sort out, but will look into trying it out

**Comment by kszatan** [comment]
I wrote a material plugin to handle Phabricator staging areas. Give it a try if you're interested:
https://github.com/kszatan/gocd-phabricator-staging-material
It's still not 100% ready but it works and is ready to receive some critis/suggestions.
